README.md
# md-mermaid-renderer
Renders all mermaid diagrams from markdown document


CHANGELOG.md
# Change Log

All notable changes to the "untill-highlighter" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

test/extension.test.js
/* global suite, test */

//
// Note: This example test is leveraging the Mocha test framework.
// Please refer to their documentation on https://mochajs.org/ for help.
//

// The module 'assert' provides assertion methods from node
const assert = require('assert');

// You can import and use all API from the 'vscode' module
// as well as import your extension to test it
const vscode = require('vscode');
//const myExtension = require('../extension');

// Defines a Mocha test suite to group tests of similar kind together
suite("Extension Tests", function() {

	// Defines a Mocha unit test
	test("Something 1", function() {
		assert.equal(-1, [1, 2, 3].indexOf(5));
		assert.equal(-1, [1, 2, 3].indexOf(0));
	});
});


src/extension.js
const vscode = require('vscode');
const renderMermaid = require('./mermaid_renderer');

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
const activate = (context) => {
	// Use the console to output diagnostic information (console.log) and errors (console.error)
	// This line of code will only be executed once when your extension is activated

	var cfg = vscode.workspace.getConfiguration('untillpro');

	// The commands has been defined in the package.json file
	// Now provide the implementation of the command with  registerCommand
	// The commandId parameter must match the command field in package.json
	let commands = [
		vscode.commands.registerCommand('markdown-mermaid-renderer.renderMermaid',  function() {
			renderMermaid();
		}),
	]

	context.subscriptions.concat(commands);
};

exports.activate = activate;

// this method is called when your extension is deactivated
const deactivate = () => {
};

exports.deactivate = deactivate;


webpack.config.js
'use strict';

const path = require('path');

/**@type {import('webpack').Configuration}*/
const config = {
  target: 'node', // vscode extensions run in a Node.js-context 📖 -> https://webpack.js.org/configuration/node/
  entry: './src/extension.js', // the entry point of this extension, 📖 -> https://webpack.js.org/configuration/entry-context/
  output: {
    // the bundle is stored in the 'dist' folder (check package.json), 📖 -> https://webpack.js.org/configuration/output/
    path: path.resolve(__dirname, 'dist'),
    filename: 'extension.js',
    libraryTarget: 'commonjs2',
    devtoolModuleFilenameTemplate: '../[resource-path]'
  },
  devtool: 'source-map',
  externals: {
    vscode: 'commonjs vscode', // the vscode-module is created on-the-fly and must be excluded. Add other modules that cannot be webpack'ed, 📖 -> https://webpack.js.org/configuration/externals/
    mime: 'mime'
  },
  resolve: {
    // support reading TypeScript and JavaScript files, 📖 -> https://github.com/TypeStrong/ts-loader
    extensions: ['.ts', '.js']
  },
  module: {
    rules: [
      {
        test: /\.ts$/,
        exclude: /node_modules/,
        use: [
          {
            loader: 'ts-loader'
          }
        ]
      }
    ]
  }
};

module.exports = config;

src/mermaid_rendere.js
#!/usr/bin/env node
'use strict';
const vscode = require('vscode');
const chalk = require('chalk');
const fs = require('fs');
const os = require('os');
const path = require('path');
const puppeteer = require('puppeteer');

const error = message => {
    console.log(chalk.red(`\n${message}\n`));
    process.exit(1);
};

const checkConfigFile = file => {
    if (!fs.existsSync(file)) {
        error(`Configuration file "${file}" doesn't exist`);
    }
};

const _asyncToGenerator = (fn) => {
    return function () {
        var gen = fn.apply(this, arguments);
        return new Promise(function (resolve, reject) {
            function step(key, arg) {
                try {
                    var info = gen[key](arg);
                    var value = info.value;
                } catch (error) {
                    reject(error);
                    return;
                }
                if (info.done) {
                    resolve(value);
                } else {
                    return Promise.resolve(value).then(function (value) {
                        step("next", value);
                    }, function (err) {
                        step("throw", err);
                    });
                }
            }
            return step("next");
        });
    };
}

const renderMermaidChunk = (i = null, o = null, t = 'default', w = 800, h = 600, bgcolor = null, cfg, css, pupCfg) => {
    let theme = t;
    let width = w;
    let height = h;
    let input = i;
    let output = o;
    let backgroundColor = bgcolor;
    let configFile = cfg;
    let cssFile = css;
    let puppeteerConfigFile = pupCfg;

    // check input file
    if (!input) {
        error('Please specify input file: -i <input>');
    }

    if (!fs.existsSync(input)) {
        error(`Input file "${input}" doesn't exist`);
    }

    if (!output) {
        output = input + '.svg';
    }

    // check output file

    if (!/\.(?:svg|png|pdf)$/.test(output)) {
        error(`Output file must end with ".svg", ".png" or ".pdf"`);
    }
    const outputDir = path.dirname(output);
    if (!fs.existsSync(outputDir)) {
        error(`Output directory "${outputDir}/" doesn't exist`);
    }

    // check config files
    let mermaidConfig = {
        theme
    };

    if (configFile) {
        checkConfigFile(configFile);
        mermaidConfig = Object.assign(mermaidConfig, JSON.parse(fs.readFileSync(configFile, 'utf-8')));
    }
    let puppeteerConfig = {};
    if (puppeteerConfigFile) {
        checkConfigFile(puppeteerConfigFile);
        puppeteerConfig = JSON.parse(fs.readFileSync(puppeteerConfigFile, 'utf-8'));
    }

    // check cssFile
    let myCSS;
    if (cssFile) {
        if (!fs.existsSync(cssFile)) {
            error(`CSS file "${cssFile}" doesn't exist`);
        }
        myCSS = fs.readFileSync(cssFile, 'utf-8');
    }

    // normalize args
    width = parseInt(width);
    height = parseInt(height);
    backgroundColor = backgroundColor || 'white';

    _asyncToGenerator(function* () {
        const browser = yield puppeteer.launch(puppeteerConfig);
        const page = yield browser.newPage();

        page.setViewport({
            width,
            height
        });
        yield page.goto(`file://${path.join(__dirname, 'index.html')}`);
        yield page.evaluate(`document.body.style.background = '${backgroundColor}'`);
        const definition = fs.readFileSync(input, 'utf-8');

        yield page.$eval('#container', function (container, definition, mermaidConfig, myCSS) {
            container.innerHTML = definition;
            window.mermaid.initialize(mermaidConfig);
    
            if (myCSS) {
                const head = window.document.head || window.document.getElementsByTagName('head')[0];
                const style = document.createElement('style');
                style.type = 'text/css';
                if (style.styleSheet) {
                    style.styleSheet.cssText = myCSS;
                } else {
                    style.appendChild(document.createTextNode(myCSS));
                }
                head.appendChild(style);
            }
    
            window.mermaid.init(undefined, container);
        }, definition, mermaidConfig, myCSS);
        
        if (output.endsWith('svg')) {
            const svg = yield page.$eval('#container', function (container) {
                return container.innerHTML;
            });
            fs.writeFileSync(output, svg);
        } else if (output.endsWith('png')) {
            const clip = yield page.$eval('svg', function (svg) {
                const react = svg.getBoundingClientRect();
                return {
                    x: react.left,
                    y: react.top,
                    width: react.width,
                    height: react.height
                };
            });
            yield page.screenshot({
                path: output,
                clip,
                omitBackground: backgroundColor === 'transparent'
            });
        } else {
            // pdf
            yield page.pdf({
                path: output,
                printBackground: backgroundColor !== 'transparent'
            });
        }

        fs.unlinkSync(input);
    
        browser.close();
    })();
};

const getOutputSpecifiedFileName = (filename) => {
    const activeEditor = vscode.window.activeTextEditor;
     
    const editorPath = activeEditor ? activeEditor.document.uri.fsPath : undefined
    const folderPath = path.dirname(editorPath);

    return `${folderPath}/${filename}`;
}

const getOutputFileName = (num) => {
    const n = num || 0;

    const activeEditor = vscode.window.activeTextEditor;
     
    const editorPath = activeEditor ? activeEditor.document.uri.fsPath : undefined
    const folderPath = path.dirname(editorPath);
    const filename = path.parse(editorPath).name;

    return `${folderPath}/${filename}_mrmd_${n}.png`;
}

const getInputFileName = (num) => {
    const n = num || 0;

    const activeEditor = vscode.window.activeTextEditor;
     
    const editorPath = activeEditor ? activeEditor.document.uri.fsPath : undefined
    const folderPath = path.dirname(editorPath);
    const filename = path.parse(editorPath).name;
    
    return `${folderPath}/${filename}_input_${n}.md`;
}

const renderMermaid = () => {
    //TODO проверки

    const activeEditor = vscode.window.activeTextEditor;
    const linesCount = parseInt(activeEditor.document.lineCount, 10);

    const startReg = /^```mermaid(?:\s+(\w+\.(?:svg|png|pdf)))?\s*$/;
    const endReg = /^```\s*$/;

    let write = false;
    let tmp_str = '';

    let inputFile = null;
    let outputFile = null;

    let graphsCount = 0;

    for (var i = 0; i < linesCount; i++) {
        const line = activeEditor.document.lineAt(i);

        const resEnd = line.text.match(endReg);
        if (resEnd) {
            write = false;

            inputFile = getInputFileName(graphsCount);
            fs.writeFileSync(inputFile, tmp_str);

            renderMermaidChunk(inputFile, outputFile);

            vscode.window.showInformationMessage(`${outputFile} was successfuly generated`);
        }

        if (write) tmp_str += `${line.text}${os.EOL}`;

        const resStart = line.text.match(startReg)

        if (resStart) {
            graphsCount++;

            write = true;

            tmp_str = '';

            if (resStart[1] && typeof resStart[1] === 'string') {
                outputFile = getOutputSpecifiedFileName(resStart[1]);
            } else {
                outputFile = getOutputFileName(graphsCount);
            }
        }
    }
}

module.exports = renderMermaid;

src/mermaid_renderer.js
'use strict';
const vscode = require('vscode');
const fs = require('fs');
const os = require('os');
const path = require('path');
const puppeteer = require('puppeteer');
const tmp = require('tmp');

const error = message => {
    console.log(`\n${message}\n`);
    process.exit(1);
};

const checkConfigFile = file => {
    if (!fs.existsSync(file)) {
        error(`Configuration file "${file}" doesn't exist`);
    }
};

const _asyncToGenerator = (fn) => {
    return function () {
        var gen = fn.apply(this, arguments);
        return new Promise(function (resolve, reject) {
            function step(key, arg) {
                try {
                    var info = gen[key](arg);
                    var value = info.value;
                } catch (error) {
                    reject(error);
                    return;
                }
                if (info.done) {
                    resolve(value);
                } else {
                    return Promise.resolve(value).then(function (value) {
                        step("next", value);
                    }, function (err) {
                        step("throw", err);
                    });
                }
            }
            return step("next");
        });
    };
}

const renderMermaidChunk = (i = null, o = null, t = 'default', w = 800, h = 600, bgcolor = null, cfg, css, pupCfg) => {
    let theme = t;
    let width = w;
    let height = h;
    let input = i;
    let output = o;
    let backgroundColor = bgcolor;
    let configFile = cfg;
    let cssFile = css;
    let puppeteerConfigFile = pupCfg;

    // check input file
    if (!input) {
        error('Please specify input file: -i <input>');
    }

    if (!fs.existsSync(input)) {
        error(`Input file "${input}" doesn't exist`);
    }

    if (!output) {
        output = input + '.svg';
    }

    // check output file

    if (!/\.(?:svg|png|pdf)$/.test(output)) {
        error(`Output file must end with ".svg", ".png" or ".pdf"`);
    }
    const outputDir = path.dirname(output);
    if (!fs.existsSync(outputDir)) {
        error(`Output directory "${outputDir}/" doesn't exist`);
    }

    // check config files
    let mermaidConfig = {
        theme
    };

    if (configFile) {
        checkConfigFile(configFile);
        mermaidConfig = Object.assign(mermaidConfig, JSON.parse(fs.readFileSync(configFile, 'utf-8')));
    }
    let puppeteerConfig = {};
    if (puppeteerConfigFile) {
        checkConfigFile(puppeteerConfigFile);
        puppeteerConfig = JSON.parse(fs.readFileSync(puppeteerConfigFile, 'utf-8'));
    }

    // check cssFile
    let myCSS;
    if (cssFile) {
        if (!fs.existsSync(cssFile)) {
            error(`CSS file "${cssFile}" doesn't exist`);
        }
        myCSS = fs.readFileSync(cssFile, 'utf-8');
    }

    // normalize args
    width = parseInt(width);
    height = parseInt(height);
    backgroundColor = backgroundColor || 'white';

    return _asyncToGenerator(function* () {
        const browser = yield puppeteer.launch(puppeteerConfig);
        const page = yield browser.newPage();

        page.setViewport({
            width,
            height
        });
        yield page.goto(`file://${path.join(__dirname, 'index.html')}`);
        yield page.evaluate(`document.body.style.background = '${backgroundColor}'`);
        const definition = fs.readFileSync(input, 'utf-8');

        yield page.$eval('#container', function (container, definition, mermaidConfig, myCSS) {
            container.innerHTML = definition;
            window.mermaid.initialize(mermaidConfig);
    
            if (myCSS) {
                const head = window.document.head || window.document.getElementsByTagName('head')[0];
                const style = document.createElement('style');
                style.type = 'text/css';
                if (style.styleSheet) {
                    style.styleSheet.cssText = myCSS;
                } else {
                    style.appendChild(document.createTextNode(myCSS));
                }
                head.appendChild(style);
            }
    
            window.mermaid.init(undefined, container);
        }, definition, mermaidConfig, myCSS);
        
        if (output.endsWith('svg')) {
            const svg = yield page.$eval('#container', function (container) {
                return container.innerHTML;
            });
            fs.writeFileSync(output, svg);
        } else if (output.endsWith('png')) {
            const clip = yield page.$eval('svg', function (svg) {
                const react = svg.getBoundingClientRect();
                return {
                    x: react.left,
                    y: react.top,
                    width: react.width,
                    height: react.height
                };
            });
            yield page.screenshot({
                path: output,
                clip,
                omitBackground: backgroundColor === 'transparent'
            });
        } else {
            // pdf
            yield page.pdf({
                path: output,
                printBackground: backgroundColor !== 'transparent'
            });
        }

        fs.unlinkSync(input);
    
        browser.close();
    })();
};

const getOutputSpecifiedFileName = (filename, folderPath) => {
    return `${folderPath}/${filename}`;
}

const getOutputFileName = (num, folderPath) => {
    const n = num || 0;

    const activeEditor = vscode.window.activeTextEditor;
     
    const editorPath = activeEditor ? activeEditor.document.uri.fsPath : undefined
    const filename = path.parse(editorPath).name;

    return `${folderPath}/${filename}_mrmd_${n}.png`;
}

const getInputFileName = (num, folderPath) => {
    const n = num || 0;
    
    const activeEditor = vscode.window.activeTextEditor;
     
    const editorPath = activeEditor ? activeEditor.document.uri.fsPath : undefined
    const filename = path.parse(editorPath).name;
    
    return `${folderPath}/${filename}_input_${n}.md`;
}

const renderMermaid = () => {
    //TODO проверки

    const activeEditor = vscode.window.activeTextEditor;
    const editorPath = activeEditor ? activeEditor.document.uri.fsPath : undefined;
    const linesCount = parseInt(activeEditor.document.lineCount, 10);

    const startReg = /^```mermaid(?:\s+(.+\.(?:svg|png|pdf)))?\s*$/i;
    const endReg = /^```\s*$/;

    let write = false;
    let tmp_str = '';

    let inputFile = null;
    let outputFile = null;

    const currentFolder = path.dirname(editorPath);
    const tmpObj = tmp.dirSync()
    const tempFolder = tmpObj.name;

    let graphsCount = 0;

    const promises = [];

    for (var i = 0; i < linesCount; i++) {
        const line = activeEditor.document.lineAt(i);

        const resEnd = line.text.match(endReg);

        if (resEnd && write) {
            write = false;

            inputFile = getInputFileName(graphsCount, tempFolder);
            fs.writeFileSync(inputFile, tmp_str);

            const p = renderMermaidChunk(inputFile, outputFile)
                .then(() => {
                    vscode.window.showInformationMessage(`${outputFile} was successfuly generated`);
                })
                .catch((e) => {
                    vscode.window.showErrorMessage(e.message);
                    fs.unlink(inputFile);
                });

            promises.push(p);
        }

        if (write) tmp_str += `${line.text}${os.EOL}`;

        const resStart = line.text.match(startReg)

        if (resStart) {
            graphsCount++;
            write = true;
            tmp_str = '';

            if (resStart[1] && typeof resStart[1] === 'string') {
                outputFile = getOutputSpecifiedFileName(resStart[1], currentFolder);
            } else {
                outputFile = getOutputFileName(graphsCount, currentFolder);
            }
        }
    }

    if (promises && promises.length > 0) {
        Promise.all(promises).then(() => {
            tmpObj.removeCallback();
        });
    }
}

module.exports = renderMermaid;